In [ ]:
# Install required packages (auto-skipped if already installed)
import importlib
if importlib.util.find_spec('qiskit') is None:
    !pip install -q qiskit qiskit-aer qiskit-ibm-runtime pylatexenc gem-suite matplotlib
else:
    print("\u2713 Packages already installed")

# To run on real quantum hardware, uncomment and fill in your credentials:
# from qiskit_ibm_runtime import QiskitRuntimeService
# QiskitRuntimeService.save_account(
#     channel="ibm_quantum_platform",
#     token="<your-api-key>",
#     # instance="<IBM Cloud CRN or instance name>",  # optional
#     set_as_default=True,
#     overwrite=True,
# )

# 西森相転移

import TutorialFeedback from '@site/src/components/TutorialFeedback';





*使用時間の目安: Heron r2 プロセッサで約3分（注: これは目安です。実際の実行時間は異なる場合があります。）*
## 学習成果
このチュートリアルを通じて、ユーザーは以下の成果を期待できます：
- 西森相転移とは何か、そしてそれがランダムボンドイジングモデルにおける長距離エンタングルメントの出現としてどのように現れるかを理解する。
- 回路中間測定と定数深さ回路を使用して、量子ハードウェア上で*測定によるエンタングルメント生成*（GEM）プロトコルを実装する。
- 実験データから2点相関と磁化の正規化分散を抽出することで、相転移を特性評価する。

## 前提条件
このチュートリアルを進める前に、以下のトピックに精通していることを推奨します：
- [Qubitの測定](/guides/measure-qubits)ガイド、特にGEMプロトコルが依存する回路中間測定に関するセクション。
- [Qiskit Aerプリミティブを使用した完全・ノイズありシミュレーション](/guides/simulate-with-qiskit-aer)（小規模セクションはこの方法で実行されます）。
- [動的回路による長距離エンタングルメント](/tutorials/long-range-entanglement)（同じ測定ベースのエンタングルメントパラダイムを使用するコンパニオンチュートリアル）。
- [ヘビーヘックス格子](https://www.ibm.com/quantum/blog/heavy-hex-lattice)（プラケット格子が構築されるIBM&reg;ハードウェアトポロジー）。
## 背景
このチュートリアルでは、量子プロセッサ上で西森相転移を実現する方法を示します。この実験は、もともと[*Realizing the Nishimori transition across the error threshold for constant-depth quantum circuits*](https://arxiv.org/abs/2309.02863)で説明されたものです。

西森相転移とは、ランダムボンドイジングモデルにおける短距離秩序相と長距離秩序相の間の転移を指します。量子コンピュータ上では、長距離秩序相はデバイス全体にわたってQubitがエンタングルされた状態として現れます。この高度にエンタングルされた状態は、*測定によるエンタングルメント生成*（GEM）プロトコルを用いて準備されます。GEMプロトコルは、回路中間測定を利用することで、定数深さの回路のみでデバイス全体のQubitをエンタングルすることができます。このチュートリアルでは、[GEM Suite](https://github.com/qiskit-community/gem-suite)ソフトウェアパッケージのGEMプロトコル実装を使用します。
## 要件
このチュートリアルを始める前に、以下がインストールされていることを確認してください：

- Qiskit SDK v1.0 以降（[visualization](https://docs.quantum.ibm.com/api/qiskit/visualization)サポート付き）
- Qiskit Runtime v0.22 以降（`pip install qiskit-ibm-runtime`）
- Qiskit Aer v0.14 以降（`pip install qiskit-aer`）
- GEM Suite（`pip install gem-suite`）
## セットアップ

In [1]:
import matplotlib.pyplot as plt
import warnings

from collections import defaultdict

from qiskit_ibm_runtime import QiskitRuntimeService
from qiskit_aer import AerSimulator

from qiskit.transpiler import generate_preset_pass_manager

from gem_suite import PlaquetteLattice
from gem_suite.experiments import GemExperiment

## 小規模シミュレーターの例

このセクションでは、ノイズのない `AerSimulator` を使用して完全なワークフローを実行します。プラケット格子は1つのプラケット（12 Qubit）に制限されているため、シミュレーションは小さく高速に保たれますが、GEMプロトコルのすべての部分（回路中間測定、$R_{ZZ}$ 角度スイープ、デコード、正規化分散分析）を実行します。同じワークフローは、後で複数のプラケットと実際のハードウェア上の完全な格子にスケールアップされます。

### ステップ 1: 古典的な入力を量子問題にマッピングする

GEMプロトコルは、格子で記述されるQubit接続性を持つ量子プロセッサ上で動作します。現在のIBM Quantum&reg;プロセッサは[ヘビーヘックス格子](https://www.ibm.com/quantum/blog/heavy-hex-lattice)を使用しています。プロセッサのQubitは、格子の単位胞のどこに位置するかに基づいて*プラケット*にグループ分けされます。1つのQubitが複数の単位胞に含まれる場合があるため、プラケットは互いに素ではありません。ヘビーヘックス格子では、1つのプラケットに12個のQubitが含まれます。プラケット自体も格子を形成しており、2つのプラケットがQubitを共有している場合に接続されます。ヘビーヘックス格子では、隣接するプラケットは3個のQubitを共有します。

GEM Suiteソフトウェアパッケージにおいて、GEMプロトコルを実装するための基本クラスは `PlaquetteLattice` です。これはプラケットの格子を表現するもので（ヘビーヘックス格子とは異なります）、Qubitの結合マップから初期化できます。現在、ヘビーヘックス結合マップのみがサポートされています。

以下のコードセルでは、量子処理ユニット（QPU）の結合マップからプラケット格子を初期化します。プラケット格子は必ずしもハードウェア全体を網羅するわけではありません。たとえば、`ibm_torino` は合計133個のQubitを持っていますが、デバイスに収まる最大のプラケット格子は125個のQubitのみを使用し、合計18個のプラケットで構成されます。`ibm_pittsburgh`（156 Qubit）も同様に144個のQubitを21個のプラケットに収めます。同じパターンは、異なるQubit数を持つ他のヘビーヘックスQPUにも当てはまります。

In [ ]:
# QiskitRuntimeService.save_account(channel="ibm_quantum", token="<YOUR_API_KEY>", overwrite=True,
# set_as_default=True)
service = QiskitRuntimeService()
backend = service.least_busy(
    operational=True, simulator=False, min_num_qubits=127
)
aer_backend = AerSimulator.from_backend(backend)
plaquette_lattice = PlaquetteLattice.from_coupling_map(backend.coupling_map)

print(f"Number of qubits in backend: {backend.num_qubits}")
print(
    f"Number of qubits in plaquette lattice: {len(list(plaquette_lattice.qubits()))}"
)
print(f"Number of plaquettes: {len(list(plaquette_lattice.plaquettes()))}")

プラケット格子のグラフ表現の図を生成することで、プラケット格子を可視化できます。図中では、プラケットはラベル付きの六角形で表され、2つのプラケットがQubitを共有している場合に辺で結ばれます。

In [3]:
plaquette_lattice.draw_plaquettes()

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/625882a4-faeb-4d96-b441-c989f43c4dea-0.avif" alt="Output of the previous code cell" />

![Output of the previous code cell](../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/625882a4-faeb-4d96-b441-c989f43c4dea-0.avif)

`plaquettes` メソッドを使用して、個々のプラケットに含まれるQubitなどの情報を取得できます。

In [4]:
# Get a list of the plaquettes
plaquettes = list(plaquette_lattice.plaquettes())
# Display information about plaquette 0
plaquettes[0]

PyPlaquette(index=0, qubits=[3, 4, 5, 6, 7, 16, 17, 23, 24, 25, 26, 27], neighbors=[4, 3, 1])

You can also produce a diagram of the underlying qubits that form the plaquette lattice.

In [5]:
plaquette_lattice.draw_qubits()

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/a19d63ce-3572-4081-a008-c1332fbbe303-0.avif" alt="Output of the previous code cell" />

プラケット格子を構成する基盤となるQubitの図も作成できます。

In [6]:
# Filter the plaquette lattice down to a single plaquette (12 qubits)
# so the AerSimulator run stays fast. The full lattice is used later
# in the large-scale hardware example.
gem_exp = GemExperiment(plaquette_lattice.filter([9]), backend=aer_backend)

# visualize the plaquette lattice after filtering
plaquette_lattice.filter([9]).draw_qubits()

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/02357c6e-5c83-4ac0-811d-22602d9f33d5-0.avif" alt="Output of the previous code cell" />

![Output of the previous code cell](../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/a19d63ce-3572-4081-a008-c1332fbbe303-0.avif)

Qubitのラベルと接続を示す辺に加えて、図にはGEMプロトコルに関連する3つの追加情報が含まれています：
- 各Qubitは塗りつぶし（グレー）または塗りつぶしなしのいずれかです。塗りつぶされたQubitはイジングモデルのサイトを表す「サイト」Qubitであり、塗りつぶされていないQubitはサイトQubit間の相互作用を媒介する「ボンド」Qubitです。
- 各サイトQubitには (A) または (B) のラベルが付けられており、GEMプロトコルにおいてサイトQubitが果たす2つの役割のいずれかを示しています（役割については後述します）。
- 各辺は6色のうちの1色で色分けされており、辺を6つのグループに分割しています。この分割は、2QubitゲートをどのようにParalelizeするか、およびノイズのある量子プロセッサ上で異なる量のエラーを引き起こす可能性のある異なるスケジューリングパターンを決定します。各グループ内の辺は互いに素であるため、それらの辺に対して2Qubitゲートの層を同時に適用できます。実際には、6色を2色ずつ3つのグループに分割し、各2色グループの和集合が依然として互いに素になるようにすることが可能です。したがって、すべての辺を活性化するために必要な2Qubitゲートの層は3層のみです。6色をこのように分割する方法は12通りあり、それぞれの分割が異なる3層のGate・スケジュールを生成します。

プラケット格子を作成したので、次のステップは `GemExperiment` オブジェクトを初期化することです。プラケット格子と実験を実行するBackendの両方を渡します。`GemExperiment` クラスは、回路の生成、ジョブの送信、データの分析を含むGEMプロトコルの実際の実装を管理します。以下のコードセルでは、プラケット格子を1つのプラケット（12 Qubit）のみに制限して実験クラスを初期化し、シミュレーションを小さく高速に保ちます。完全なプラケット格子は、後で実際のハードウェアにスケールアップする際に使用します。

In [7]:
circuits = gem_exp.circuits()
print(f"Total number of circuits: {len(circuits)}")

Total number of circuits: 252


![Output of the previous code cell](../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/02357c6e-5c83-4ac0-811d-22602d9f33d5-0.avif)

GEMプロトコルのCircuitは、以下の手順で構築されます：
1. すべてのQubitにアダマールGateを適用して、全 $|+\rangle$ 状態を準備します。
2. 接続されたすべてのQubitペアの間に $R_{ZZ}$ Gateを適用します。これは3層のGateで実現できます。各 $R_{ZZ}$ GateはサイトQubitとボンドQubitに作用します。サイトQubitのラベルが (B) の場合、角度は $\frac{\pi}{2}$ に固定されます。サイトQubitのラベルが (A) の場合、角度は可変であり、異なるCircuitを生成します。デフォルトでは、角度の範囲は $0$ から $\frac{\pi}{2}$ までの両端を含む等間隔の21点に設定されています。
3. 各ボンドQubitをパウリ $X$ 基底で測定します。Qubitはパウリ $Z$ 基底で測定されるため、測定前にアダマールGateを適用することでこれを実現できます。

このチュートリアルの冒頭で引用した論文では、$R_{ZZ}$ 角度に異なる規約を使用しており、このチュートリアルで使用する規約とは2倍の係数が異なることに注意してください。

ステップ3では、ボンドQubitのみが測定されます。サイトQubitがどのような状態に残されるかを理解するために、ステップ2でサイトQubit (A) に適用される $R_{ZZ}$ 角度が $\frac{\pi}{2}$ に等しい場合を考えると有益です。この場合、サイトQubitはGHZ状態に類似した高度にエンタングルされた状態に置かれます、

$$
\lvert \text{GHZ} \rangle = \lvert 00 \cdots 00 \rangle + \lvert 11 \cdots 11 \rangle.
$$

測定結果のランダム性により、サイトQubitの実際の状態は、たとえば $\lvert 00110 \rangle + \lvert 11001 \rangle$ のような、長距離秩序を持つ別の状態になる場合があります。しかし、測定結果に基づくデコード操作を適用することで、GHZ状態を回復できます。$R_{ZZ}$ 角度を $\frac{\pi}{2}$ から小さくしていくと、臨界角度まで長距離秩序を回復できます。ノイズがない場合、臨界角度は約 $0.3 \pi$ です。この角度を下回ると、得られる状態は長距離エンタングルメントを示さなくなります。この長距離秩序の有無の間の転移が、西森相転移です。

上記の説明では、サイトQubitは測定されずに残され、デコード操作は量子Gateを適用することで実行できます。GEM Suiteで実装された実験では、実際にはサイトQubitも測定され、デコード操作は古典的な後処理ステップで適用されます。

上記の説明では、デコード操作はサイトQubitに量子Gateを適用して量子状態を回復することで実行できます。しかし、たとえば特性評価の目的で状態を直ちに測定することが目的の場合、サイトQubitはボンドQubitとともに測定され、デコード操作は古典的な後処理ステップで適用できます。

ステップ2の $R_{ZZ}$ 角度（デフォルトでは21個の値をスイープ）に加えて、GEMプロトコル回路は $R_{ZZ}$ Gateの3層を実装するために使用されるスケジューリングパターンにも依存します。前述のとおり、このようなスケジューリングパターンは12通りあります。したがって、実験の回路総数は $21 \times 12 = 252$ です。

実験の回路は、`GemExperiment` クラスの `circuits` メソッドを使用して生成できます。

In [8]:
# Restrict experiment to the first scheduling pattern
gem_exp.set_experiment_options(schedule_idx=0)

# There are less circuits now
circuits = gem_exp.circuits()
print(f"Total number of circuits: {len(circuits)}")

# Print the RZZ angles swept over
print(f"RZZ angles:\n{gem_exp.parameters()}")

Total number of circuits: 21
RZZ angles:
[0.         0.07853982 0.15707963 0.23561945 0.31415927 0.39269908
 0.4712389  0.54977871 0.62831853 0.70685835 0.78539816 0.86393798
 0.9424778  1.02101761 1.09955743 1.17809725 1.25663706 1.33517688
 1.41371669 1.49225651 1.57079633]


The following code cell draws a diagram of the circuit at index 5. To reduce the size of the diagram, the measurement gates at the end of the circuit are removed.

In [9]:
# Get the circuit at index 5
circuit = circuits[5]
# Remove the final measurements to ease visualization
circuit.remove_final_measurements()
# Draw the circuit
circuit.draw("mpl", fold=-1, scale=0.5)

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/fd57d483-c70b-4ad5-b309-15750ad38bac-0.avif" alt="Output of the previous code cell" />

このチュートリアルでは、単一のスケジューリングパターンのみを考慮すれば十分です。以下のコードセルでは、実験を最初のスケジューリングパターンに制限します。その結果、実験にはスイープされる各 $R_{ZZ}$ 角度に1つずつ、合計21個の回路のみが含まれます。

In [10]:
# Demonstrate setting transpile options
gem_exp.set_transpile_options(
    optimization_level=1  # This is the default optimization level
)
pass_manager = generate_preset_pass_manager(
    backend=aer_backend,
    initial_layout=list(gem_exp.physical_qubits),
    **dict(gem_exp.transpile_options),
)
transpiled = pass_manager.run(circuit)
transpiled.draw("mpl", idle_wires=False, fold=-1, scale=0.5)

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/e9b99d48-8d33-46b5-bff5-480ab1c1c1f2-0.avif" alt="Output of the previous code cell" />

### Step 3: Execute using Qiskit primitives

To execute the GEM protocol circuits on the hardware, call the `run` method of the `GemExperiment` object. You can specify the number of shots you want to sample from each circuit. The `run` method returns an [ExperimentData](https://qiskit-community.github.io/qiskit-experiments/stubs/qiskit_experiments.framework.ExperimentData.html) object which you should save to a variable. Note that the `run` method only submits jobs without waiting for them to finish, so it is a non-blocking call.

In [11]:
exp_data = gem_exp.run(shots=10_000)

以下のコードセルでは、インデックス5の回路の図を描画します。図のサイズを縮小するため、回路末尾の測定Gateは除去されています。

In [12]:
# The noiseless AerSimulator produces zero-variance UFloat objects in the
# analysis, which triggers a harmless warning from the `uncertainties`
# library. Suppress it so the output stays clean.
with warnings.catch_warnings():
    warnings.filterwarnings(
        "ignore", message="Using UFloat objects with std_dev==0"
    )
    exp_data.block_for_results()
exp_data

ExperimentData(GemExperiment, 90bf2a90-f729-4c4e-a6da-664aecb11039, job_ids=['04a7c405-47fd-46ca-aa4b-aaf7e339cfbe'], metadata=<5 items>, figure_names=['two_point_correlation.svg', 'normalized_variance.svg', 'plaquette_ops.svg', 'bond_ops.svg'])

![Output of the previous code cell](../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/fd57d483-c70b-4ad5-b309-15750ad38bac-0.avif)

### ステップ 2: 量子ハードウェア実行のための問題の最適化
量子CircuitをハードウェアでTranspileするには、通常[複数のステージ](/guides/transpiler-stages)が含まれます。一般的に、最も計算コストがかかるステージは、Qubitレイアウトの選択、ハードウェアのQubit接続性に合わせた2QubitゲートのルーティングTranspile、およびゲート数と深さを最小化するための回路最適化です。GEMプロトコルでは、ハードウェアの接続性がすでにプロトコルの設計に組み込まれているため、レイアウトとルーティングのステージは不要です。回路にはすでにQubitレイアウトがあり、2QubitゲートはすでにネイティブCircuit接続にマッピングされています。さらに、$R_{ZZ}$ 角度を変化させる際に回路の構造を保持するため、ごく基本的な回路最適化のみを実行する必要があります。

`GemExperiment` クラスは、実験の実行時に回路を透過的にTranspileします。レイアウトとルーティングのステージはデフォルトで何もしないようにオーバーライドされており、回路最適化は単一QubitGateのみを最適化するレベルで実行されます。ただし、`set_transpile_options` メソッドを使用して、オーバーライドしたり追加のオプションを渡したりすることができます。可視化のため、以下のコードセルでは前に表示した回路を手動でTranspileし、Transpile後の回路を描画します。

In [13]:
def magnetization_distribution(
    counts_dict: dict[str, int],
) -> dict[str, float]:
    """Compute magnetization distribution from counts dictionary."""
    # Construct dictionary from magnetization to count
    mag_dist = defaultdict(float)
    for bitstring, count in counts_dict.items():
        mag = bitstring.count("0") - bitstring.count("1")
        mag_dist[mag] += count
    # Normalize
    shots = sum(counts_dict.values())
    for mag in mag_dist:
        mag_dist[mag] /= shots
    return mag_dist


# Get counts dictionaries with and without decoding
data = exp_data.data()
# Get the last data point, which is at the angle for the GHZ state
raw_counts = data[-1]["counts"]
# Without decoding
site_indices = [
    i for i, q in enumerate(gem_exp.plaquettes.qubits()) if q.role == "Site"
]
site_raw_counts = defaultdict(int)
for key, val in raw_counts.items():
    site_str = "".join(key[-1 - i] for i in site_indices)
    site_raw_counts[site_str] += val
# With decoding
_, site_decoded_counts = gem_exp.plaquettes.decode_outcomes(
    raw_counts, return_counts=True
)

# Compute magnetization distribution
raw_magnetization = magnetization_distribution(site_raw_counts)
decoded_magnetization = magnetization_distribution(site_decoded_counts)

# Plot
plt.bar(*zip(*raw_magnetization.items()), label="raw")
plt.bar(*zip(*decoded_magnetization.items()), label="decoded", width=0.3)
plt.legend()
plt.xlabel("Magnetization")
plt.ylabel("Frequency")
plt.title("Magnetization distribution with and without decoding")

Text(0.5, 1.0, 'Magnetization distribution with and without decoding')

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/8ead3582-16df-4616-836c-bdce867ad6b8-1.avif" alt="Output of the previous code cell" />

![Output of the previous code cell](../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/e9b99d48-8d33-46b5-bff5-480ab1c1c1f2-0.avif)

### ステップ 3: Qiskitプリミティブを使用した実行
GEMプロトコル回路をハードウェア上で実行するには、`GemExperiment` オブジェクトの `run` メソッドを呼び出します。各回路からサンプリングするショット数を指定できます。`run` メソッドは [ExperimentData](https://qiskit-community.github.io/qiskit-experiments/stubs/qiskit_experiments.framework.ExperimentData.html) オブジェクトを返しますので、これを変数に保存してください。`run` メソッドはジョブの送信のみを行い、完了を待たないため、ノンブロッキング呼び出しであることに注意してください。

In [14]:
exp_data.figure("two_point_correlation")

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/4ecb25c8-e572-49af-a879-9943039db131-0.avif" alt="Output of the previous code cell" />

結果を待つには、`ExperimentData` オブジェクトの `block_for_results` メソッドを呼び出します。この呼び出しにより、ジョブが完了するまでインタープリタはブロックされます。

In [15]:
exp_data.figure("normalized_variance")

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/2b351d68-3924-445a-94ef-047b16214e8a-0.avif" alt="Output of the previous code cell" />

## Large-scale hardware example

Having validated the protocol on a simulator, you can now scale up the experiment and run it on the real quantum hardware backend selected in the [Setup](#setup) section. This example uses two larger problem sizes:

- **Six plaquettes (~49 qubits)**: a mid-size run that already shows the rightward shift of the critical point under hardware noise.
- **The full plaquette lattice**: every plaquette the device's heavy-hex topology supports (for example, 18 plaquettes / 125 qubits on `ibm_torino` or 21 plaquettes / 144 qubits on `ibm_pittsburgh`), entangling qubits across the entire device with constant-depth circuits.

The single code cell below is self-contained: it builds the plaquette lattice from the backend's coupling map and runs both experiments, so this section can be executed after the [Setup](#setup) cells without first running the small-scale section.

In [ ]:
# -------------------------Step 1-------------------------
# Initialize the runtime service, pick a real quantum hardware backend,
# and build the plaquette lattice from its coupling map. This is repeated
# from the small-scale example so this cell can run standalone after the
# Setup section. The full plaquette lattice is the "large-scale" target;
# a six-plaquette subset (range(3, 9)) is also used to show an intermediate
# scaling step.
service = QiskitRuntimeService()
backend = service.least_busy(
    operational=True, simulator=False, min_num_qubits=127
)
plaquette_lattice = PlaquetteLattice.from_coupling_map(backend.coupling_map)

# Build a GemExperiment for the full plaquette lattice and one for the
# six-plaquette subset, each restricted to a single scheduling pattern so
# the experiment has one circuit per RZZ angle (21 circuits total).
gem_exp_full = GemExperiment(plaquette_lattice, backend=backend)
gem_exp_full.set_experiment_options(schedule_idx=0)
gem_exp_6 = GemExperiment(
    plaquette_lattice.filter(range(3, 9)), backend=backend
)
gem_exp_6.set_experiment_options(schedule_idx=0)

circuits = gem_exp_full.circuits()
print(f"Total number of circuits (full lattice): {len(circuits)}")

# -------------------------Step 2-------------------------
# GemExperiment transpiles internally for the target backend: the layout
# and routing stages are overridden because the plaquette lattice already
# matches the hardware connectivity, and optimization is restricted so the
# RZZ angle structure is preserved. The code below manually transpiles one
# circuit from the six-plaquette experiment with the same settings this
# experiment will use, and draws it for inspection. (The full-lattice
# transpiled circuit has too many qubits to visualize cleanly, so the
# six-plaquette circuit is used here as a representative example.)
gem_exp_6.set_transpile_options(optimization_level=1)
circuits_6 = gem_exp_6.circuits()
pass_manager = generate_preset_pass_manager(
    backend=backend,
    initial_layout=list(gem_exp_6.physical_qubits),
    **dict(gem_exp_6.transpile_options),
)
transpiled = pass_manager.run(circuits_6[5])
display(transpiled.draw("mpl", idle_wires=False, fold=-1, scale=0.5))

# -------------------------Step 3-------------------------
# Run both problem sizes on real hardware:
#   1. Six plaquettes (~49 qubits) — an intermediate scale-up.
#   2. The full plaquette lattice — every plaquette the device supports.
exp_data_6 = gem_exp_6.run(shots=10_000, job_tags=["TUT_NPT"])
exp_data_full = gem_exp_full.run(shots=10_000, job_tags=["TUT_NPT"])
exp_data_6.block_for_results()
exp_data_full.block_for_results()

# -------------------------Step 4-------------------------
# Plot the normalized variance at each scale. The peak marks the critical
# point of the Nishimori transition; as the system grows, hardware noise
# shifts the peak rightward.
display(exp_data_6.figure("normalized_variance"))
exp_data_full.figure("normalized_variance")

Total number of circuits (full lattice): 21


<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/08581c09-a6a5-4a56-9fc4-abf22b063c6a-1.avif" alt="Output of the previous code cell" />

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/08581c09-a6a5-4a56-9fc4-abf22b063c6a-2.avif" alt="Output of the previous code cell" />

<Image src="../docs/images/tutorials/nishimori-phase-transition/extracted-outputs/08581c09-a6a5-4a56-9fc4-abf22b063c6a-3.avif" alt="Output of the previous code cell" />

### ステップ 4: 後処理と所望の古典形式での結果の返却
$R_{ZZ}$ 角度が $\frac{\pi}{2}$ の場合、ノイズがなければデコードされた状態はGHZ状態になります。GHZ状態の長距離秩序は、測定されたビット列の磁化をプロットすることで可視化できます。磁化 $M$ は、単一QubitのパウリZ演算子の和として定義されます。
$$
M = \sum_{j=1}^N Z_j,
$$
ここで $N$ はサイトQubitの数です。ビット列に対するその値は、0の数と1の数の差に等しくなります。GHZ状態を測定すると、全ゼロ状態または全1状態が等確率で得られるため、磁化は半分の確率で $+N$、もう半分の確率で $-N$ となります。ノイズによるエラーが存在する場合、他の値も現れますが、ノイズが大きすぎなければ、分布は依然として $+N$ と $-N$ の近くにピークを持ちます。

デコード前の生のビット列については、ノイズがない場合、磁化の分布は一様ランダムなビット列の分布と等価になります。

以下のコードセルでは、$R_{ZZ}$ 角度が $\frac{\pi}{2}$ の場合の生のビット列とデコードされたビット列の磁化をプロットします。